### PR TITLE
Use closefrom(2) if available in the OS

### PR DESCRIPTION
**Pull request by csjperon**
Currently, when the neverbleed process is forked we query `_SC_OPEN_MAX` and
iterate through every file descriptor above `STDERR` (skipping listen and
close notify file descriptors) and call the `close(2)` syscall on the fd.
Each syscall is a context switch, and depending on the audit/debug state of
the system during this time, roughly 231,000 syscalls will be executed.
On security audited, traced or otherwise hooked (kprobes, ebpf, perf)
processes, this will generate 231,000 records being committed to disk.

This change introduces an optimization which determines the highest file
descriptor value currently in the process, uses the standard `close(2)` syscall
for anything less than (excluding stdout, stderr, and stdin) then executes
`closefrom(2)` on the highest fd plus one. This substantially decreases the number
of context switches during the bootstrap phase of the neverbleed process (from
231,000 to roughly 9. This change also includes the necessary cmake changes to
detect the functionality. Tested and verified on FreeBSD:

```
 [..]
 52544: close(3)                                  = 0 (0x0)
 52544: close(4)                                  = 0 (0x0)
 52544: close(5)                                  = 0 (0x0)
 52544: close(6)                                  = 0 (0x0)
 52544: close(7)                                  = 0 (0x0)
 52544: close(8)                                  = 0 (0x0)
 52544: close(9)                                  = 0 (0x0)
 52544: close(11)                                 ERR#9 'Bad file descriptor'
 52544: closefrom(13)                             = 0 (0x0)
 [..]
```

### CODE REVIEW

File: deps/neverbleed/neverbleed.c [hunk 1]
```diff
@@ -1346,20 +1346,51 @@ static void *daemon_conn_thread(void *_sock_fd)
     return NULL;
 }
 
+
+static void neverbleed_cleanup_fds(int listen_fd, int close_notify_fd)
```
Comment (kazuho) [inline deps/neverbleed/neverbleed.c:1350]: I think we can remove the prefix ("neverbleed_") and call the function "cleanup_fds" here> It is a static function, we do not need to be concerned of name collision.

File: deps/neverbleed/neverbleed.c [hunk 2]
```diff
@@ -1346,20 +1346,51 @@ static void *daemon_conn_thread(void *_sock_fd)
...
+{
+#ifdef LIBC_HAS_CLOSEFROM
```
Comment (kazuho) [inline deps/neverbleed/neverbleed.c:1352]: I appreciate having detection of the libc function and deciding based on that, though the restriction here is that this is a change to neverbleed that does not even have a CMakeFile (or an autoconf file).

Would it be possible to
* check the operating system (e.g., `#if defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__)`) instead, and
* redirect this PR to https://github.com/h2o/neverbleed?

FWIW, we do that occasionally (as an example, see https://github.com/h2o/h2o/blob/master/lib/common/socket/evloop.c.h#L32-L86), and I do not mind having another.

File: deps/neverbleed/neverbleed.c (hunk 1 above)
Comment (csjperon) [reply deps/neverbleed/neverbleed.c]: Done

File: deps/neverbleed/neverbleed.c (hunk 2 above)
Comment (csjperon) [reply deps/neverbleed/neverbleed.c]: Completed. Moved to:

https://github.com/h2o/neverbleed/pull/29

**Comment by csjperon** [comment]
301 to https://github.com/h2o/neverbleed/pull/29

**Pull request closed without merging**